lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+	// could failed in the mean time the caller calls the
+	// Stop method hanging here as it will blocks to write
+	// into the channel since there is no channel reader
+	close(fh.stopCh)
+	<-fh.errorCh
+	<-fh.handlerDone
+
+	err = fh.stop()
+	return err
+}
+
+// waitServices will start the ephemeral services that handles the
+// votes for the current round and once those services finishes the
+// waitServices return without errors, if one ephemeral service face
+// a problem the waitServices will shut down all the running ephemeral
+// services and return an error, this function also returns if the
+// finalizationHandler.Stop() method is called
+func (fh *finalizationHandler) waitServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		finalizationEngineErr <- fh.finalizationEngine.Run()
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		votingRoundErr <- fh.votingRound.Run()
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err := <-votingRoundErr:
+			if err == nil {
+				votingRoundErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+
+		case err := <-finalizationEngineErr:
+			if err == nil {
+				finalizationEngineErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+		}
+
+		finish := votingRoundErr == nil && finalizationEngineErr == nil
+		if finish {
+			return nil
+		}
+	}
+}
+
+// handleVotingRound interacts with finalizationEngine service
+// executing the actions based on what it receives throuhg channel
+type handleVotingRound struct {
+	grandpaService       *Service
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	if h.engineDone == nil {
+		return nil
+	}
+
+	close(h.stopCh)
+	<-h.engineDone
+
+	return nil
+}
+
+func (h *handleVotingRound) Run() (err error) {
+	defer func() {
+		close(h.engineDone)
+		h.engineDone = nil
+	}()
+
+	start := time.Now()
+
+	logger.Debugf("starting round %d with set id %d",
+		h.grandpaService.state.round, h.grandpaService.state.setID)
+
+	for {
+		select {
+		case <-h.stopCh:
+			return nil
+
+		case action, ok := <-h.finalizationEngineCh:
+			if !ok {
+				return nil
+			}
```

Comment:
`h.finalizationEngineCh` can only be closed in the block:

```go
func (f *finalizationEngine) Stop() (err error) {
	if f.engineDone == nil {
		return nil
	}

	close(f.stopCh)
	<-f.engineDone

	close(f.actionCh)
	return nil
}
```

So the `stopCh` case should trigger first, let's just remove this otherwise it allows, if we're working on an action below, the function to exit from either of the 2 select cases (which is bad for future changes and predictability).

```suggestion
		case action := <-h.finalizationEngineCh:
```